src/metabase/mbql/normalize.cljc

```diff
@@ -639,25 +641,40 @@
 
 (defn- canonicalize-mbql-clauses
   "Walk an `mbql-query` an canonicalize non-top-level clauses like `:fk->`."
-  [mbql-query]
-  (walk/prewalk
-   (fn [x]
-     (cond
-       (map? x)
-       (m/map-vals canonicalize-mbql-clauses x)
-
-       (not (mbql-clause? x))
-       x
-
-       :else
-       (try
-         (canonicalize-mbql-clause x)
-         (catch #?(:clj Throwable :cljs js/Error) e
-           (log/error (i18n/tru "Invalid clause:") x)
-           (throw (ex-info (i18n/tru "Invalid MBQL clause: {0}" (ex-message e))
-                           {:clause x}
-                           e))))))
-   mbql-query))
+  [form]
+  (cond
+    (instance? IRecord form)
+    (reduce (fn [r x] (conj r (canonicalize-mbql-clauses x))) form form)
+
+    (map? form)
+    (m/map-vals canonicalize-mbql-clauses form)
+
+    (mbql-clause? form)
+    (let [top-canonical
+          (try
+            (canonicalize-mbql-clause form)
+            (catch #?(:clj Throwable :cljs js/Error) e
+              (log/error (i18n/tru "Invalid clause:") form)
+              (throw (ex-info (i18n/tru "Invalid MBQL clause: {0}" (ex-message e))
+                              {:clause form}
+                              e))))]
+      (if (seq top-canonical)
+        (into (conj (empty top-canonical) (first top-canonical))
+              (map canonicalize-mbql-clauses)
+              (rest top-canonical))
```

Comment:
This is assuming that the canonical MBQL clause is not a list, a lazy sequence or a map. If that's not true, this will have to get more complicated.